Cleanup Triage and enable in Beta

nev-r opened the pull request:
i'd like to enable this in beta after release.
try it out and let me know if anything needs fixing before people see it

triage tab now contains hinting about whether an item is in any loadouts, directly in the upper tab, just like the wishlist hint.
![image](https://user-images.githubusercontent.com/68782081/165000305-79fd3b37-393f-458e-8569-6687496d550e.png)

compare buttons accompany the filter buttons in "similar items" section.
![image](https://user-images.githubusercontent.com/68782081/165000319-584c804b-ccdd-4914-ad86-b8810233615a.png)

vivekhnz commented:
Looks great! I had a play around and have some comments:

1. There's a crash if you try to switch to the Triage tab when viewing an item that you don't currently own e.g. an item being sold by a vendor:
  ![dim-triage-vendor-crash](https://user-images.githubusercontent.com/17512262/165019908-3f3d5df3-829b-4de0-8430-62a997c49b5b.gif)
  You can also repro it by opening the Triage tab on a weapon in Collections that you don't currently own (filter by `-is:invault -is:inleftchar -is:inmiddlechar -is:inrightchar`)
2. The loadout icon in the Triage tab header is a nice touch, but it's not clear what it means unless you mouse over the icon (not an option on mobile). Maybe add the loadout icon to the 'In Loadouts' section header (similar to the Wish List header) so the correlation is clearer?
3. IMO, the 'Wish List Notes:' text is redundant if it's the only thing within the 'Wish List' section
   ![dim-triage-wishlist-notes-label](https://user-images.githubusercontent.com/17512262/165020737-15c3d987-255a-4640-bd91-bc9b3865d6bf.png)
4. The two-level collapse behaviour (Wish List section collapse + wish list notes text collapse) for the wishlist notes feels a bit clunky IMO. It might be worth either:
    - Removing the collapsibility from the wishlist notes. This feels like the simplest option but potentially undesirable if we have very verbose wishlist notes and want the Wish List section expanded by default. OR
    - Auto-expanding the wishlist notes text when the parent Wish List section is expanded by the user i.e.
      - Default view is Wish List section expanded & notes collapsed
      - The user can click on the wish list notes to expand the notes
      - If the user collapses the Wish List section and then expands it, the notes expand also so they can read the whole thing (which is probably the user's intention)
5. The way Adept weapons are treated under the Similar Items section is not consistent with how Compare treats them e.g.
    - I have 6 regular Deliverances + 1 Adept Deliverance and 1 regular Palindrome + 2 Adept Palindromes
    - "Triage same item" = row with item icon + name

    |Item|Sidecar compare button|Triage same item count|Triage same item compare button|
    |---|---|---|---|
    |Deliverance (regular)|7 = 6 reg + 1 adept|7|7 = 6 reg + 1 adept|
    |Deliverance (adept)|7 = 6 reg + 1 adept|Not displayed|Not displayed|
    |Palindrome (regular)|3 = 1 reg + 2 adept|3|3 = 1 reg + 2 adept|
    |Palindrome (adept)|3 = 1 reg + 2 adept|2|2|

    - One could argue that Triage is behaving more correctly than Comapre which is fine, I just think we should be consistent
6. We should hide the High Stats section on class items
7. The 'This item' label in High Stats sometimes wraps (Chrome 100, Windows 10, 100% zoom). Maybe we should
    - Increase the column width slightly to accommodate for English
    - Vertically align column headers to the bottom so the other column headers look less out of place when we inevitably wrap on other languages
    ![dim-triage-wrapping](https://user-images.githubusercontent.com/17512262/165023096-ab851fa9-babb-4438-9772-a0c2e273bc6f.png)
8. The colour of the stat percentage vs. best item text is not very readable if the percentage is greater than 100% (e.g. for vendor items that are better than anything you currently own)
   ![dim-triage-stat-pct-colour](https://user-images.githubusercontent.com/17512262/165023516-0e4183f7-90d3-4b34-bcc4-61164d48e64a.png)
9. Maybe replace the high stat table with a label ('This item doesn't have any high stats' or something) if the table is empty
10. I haven't tested this on a physical device, but it seems like the loadout edit, search and compare buttons on Triage rows are a bit small for mobile. Maybe pad these out a bit if on mobile?
11. The triage row search buttons set the current search filter, which works well on desktop when you can see your whole inventory, but isn't as useful on mobile. Maybe we should pop up the search results sheet on mobile instead?

This is really cool, looking forward to seeing it in beta. I don't think any of the above comments are deal-breakers besides the crash.

nev-r commented:
> crash if you try to switch to the Triage tab when viewing an item that you don't currently own

fixed. i have not really been considering this in terms of unowned items, so this is a great point.

> add the loadout icon to the 'In Loadouts' section header

agreed. fixed.

> 'Wish List Notes:' text is redundant if it's the only thing within the 'Wish List' section

i was feeling either way on this, but glad to hear an opinion, let's go with no notes label

> The two-level collapse behaviour (Wish List section collapse + wish list notes text collapse) for the wishlist notes feels a bit clunky IMO.

agreed. wishlists have gotten extremely wordy so going with the "1 click opens both" plan.

    The way Adept weapons are treated under the Similar Items section is not consistent with how Compare treats them e.g.
        I have 6 regular Deliverances + 1 Adept Deliverance and 1 regular Palindrome + 2 Adept Palindromes
        "Triage same item" = row with item icon + name
    Item 	Sidecar compare button 	Triage same item count 	Triage same item compare button
    Deliverance (regular) 	7 = 6 reg + 1 adept 	7 	7 = 6 reg + 1 adept
    Deliverance (adept) 	7 = 6 reg + 1 adept 	Not displayed 	Not displayed
    Palindrome (regular) 	3 = 1 reg + 2 adept 	3 	3 = 1 reg + 2 adept
    Palindrome (adept) 	3 = 1 reg + 2 adept 	2 	2
        One could argue that Triage is behaving more correctly than Compare which is fine, I just think we should be consistent

> We should hide the High Stats section on class items

agreed. done. 
    The 'This item' label in High Stats sometimes wraps (Chrome 100, Windows 10, 100% zoom). Maybe we should
        Increase the column width slightly to accommodate for English
        Vertically align column headers to the bottom so the other column headers look less out of place when we inevitably wrap on other languages
        dim-triage-wrapping

    The colour of the stat percentage vs. best item text is not very readable if the percentage is greater than 100% (e.g. for vendor items that are better than anything you currently own)
    dim-triage-stat-pct-colour

    Maybe replace the high stat table with a label ('This item doesn't have any high stats' or something) if the table is empty

    I haven't tested this on a physical device, but it seems like the loadout edit, search and compare buttons on Triage rows are a bit small for mobile. Maybe pad these out a bit if on mobile?

    The triage row search buttons set the current search filter, which works well on desktop when you can see your whole inventory, but isn't as useful on mobile. Maybe we should pop up the search results sheet on mobile instead?

nev-r commented:
thank you so much for the thorough check-up

> crash if you try to switch to the Triage tab when viewing an item that you don't currently own

fixed. i have not really been considering this in terms of unowned items, so this is a great point. i am looking around for other weirdness on this.

> add the loadout icon to the 'In Loadouts' section header

agreed. fixed.

> 'Wish List Notes:' text is redundant if it's the only thing within the 'Wish List' section

i was feeling either way on this, but glad to hear an opinion, let's go with no notes label

> The two-level collapse behaviour (Wish List section collapse + wish list notes text collapse) for the wishlist notes feels a bit clunky IMO.

agreed. wishlists have gotten extremely wordy so going with the "1 click opens both" plan.

> The way Adept weapons are treated under the Similar Items section is not consistent with how Compare treats them

making this match compare for now.

> We should hide the High Stats section on class items

agreed. done. 

> Maybe replace the high stat table with a label if the table is empty

making it disabled and --'d out

> The colour of the stat percentage vs. best item text is not very readable if the percentage is greater than 100%

fixed

> pop up the search results sheet on mobile instead?

definitely willing to follow up and do this

> The 'This item' label in High Stats sometimes wraps

investigating this

> buttons on Triage rows are a bit small for mobile

definitely willing to follow up and do this

nev-r commented:
column headers now shouldn't wrap. we'll address if that causes any problem.
several controls are now bigger on mobile,

bhollis commented:
I'm excited to try this out with real users!

robojumper commented on src/app/item-triage/ItemTriage.tsx:
```diff
@@ -55,321 +40,232 @@ export function doShowTriage(item: DimItem) {
+      <ul className={styles.loadoutList}>
+        {inLoadouts.map((l) => {
+          const edit = () => {
+            editLoadout(l, item.owner, {
```
This edits loadouts in the context of the wrong store when the item is owned by a different character, e.g. I have a "Verity's + Headstone" Warlock loadout, but when I click the "edit" icon in Triage while the Eyasluna in the loadout is owned by my Titan, it hides all Warlock items from the loadout while editing (leading to missing armor and subclass).

Not sure if the fix should be here or down in the loadout drawer, but it's not super urgent because loadouts don't seem to change their class type or drop items when saving.

nev-r replied:
fixed missing armor by editing in context of the vault. let me know if that seems likely to do something weird. it looks and save right, from simple testing

robojumper replied:
#8344

robojumper reviewed (approved):
🚀
